File: versions-utils-java/src/main/java/ru/dreamkas/semver/prerelease/PreReleaseStringId.java
package ru.dreamkas.semver.prerelease;

import org.jetbrains.annotations.NotNull;

public final class PreReleaseStringId implements PreReleaseId {
    private final String id;

    public PreReleaseStringId(@NotNull String id) {
        this.id = id;
    }

    @Override
    public int compareTo(@NotNull PreReleaseId o) {
        if (!(o instanceof PreReleaseStringId)) {
            return 1;
        }
        return id.compareToIgnoreCase(((PreReleaseStringId) o).id);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof PreReleaseStringId)) {
            return false;
        }
        PreReleaseStringId that = (PreReleaseStringId) o;
        return id.equalsIgnoreCase(that.id);
    }

    @Override
    public int hashCode() {
        return id.toLowerCase().hashCode();
    }
}

File: versions-utils/src/main/java/ru/dreamkas/semver/prerelease/PreReleaseStringId.kt
package ru.dreamkas.semver.prerelease

class PreReleaseStringId(val id: String) : PreReleaseId {
    override fun compareTo(other: PreReleaseId): Int {
        return if (other is PreReleaseNumericId) 1
        else id.compareTo((other as PreReleaseStringId).id, ignoreCase = true)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other?.javaClass != javaClass) return false

        other as PreReleaseStringId

        if (id.toLowerCase() != other.id.toLowerCase()) return false

        return true
    }

    override fun hashCode(): Int {
        return id.toLowerCase().hashCode()
    }

}